src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp

```diff
@@ -0,0 +1,143 @@
+
+/**
+ * The XGboostTreeRegressor class provides the implementation of Gradient
+ * Boosted Decision Tree Regressor as described in the XGBoost paper:
+ *
+ * @code
+ * @inproceedings{Chen:2016:XST:2939672.2939785,
+ *   author = {Chen, Tianqi and Guestrin, Carlos},
+ *   title = {{XGBoost}: A Scalable Tree Boosting System},
+ *   booktitle = {Proceedings of the 22nd ACM SIGKDD International Conference on
+ *                Knowledge Discovery and Data Mining},
+ *   series = {KDD '16},
+ *   year = {2016},
+ *   isbn = {978-1-4503-4232-2},
+ *   location = {San Francisco, California, USA},
+ *   pages = {785--794},
+ *   numpages = {10},
+ *   url = {http://doi.acm.org/10.1145/2939672.2939785},
+ *   doi = {10.1145/2939672.2939785},
+ *   acmid = {2939785},
+ *   publisher = {ACM},
+ *   address = {New York, NY, USA},
+ *   keywords = {large-scale machine learning},
+ * }
+ * @endcode
+ */
+template<typename LossFunction = SSELoss,
+         typename DimensionSelectionType =
+             mlpack::tree::MultipleRandomDimensionSelect,
+         template <typename> class NumericSplitType = XGBExactNumericSplit>
+class XGBoostTreeRegressor
+{
+ public:
+  //! Allow access to the underlying decision tree type.
+  typedef mlpack::tree::DecisionTreeRegressor<FitnessFunction,
+      NumericSplitType, mlpack::tree::AllCategoricalSplit,
+      DimensionSelectionType> DecisionTreeType;
+  /**
+   * Construct the xgboost without any training or specifying the number
+   * of trees.  Predict() will throw an exception until Train() is called.
+   */
+  XGBoostTreeRegressor();
+
+  /**
+   * Construct the xgboost forest and train.
+   */
+  template<typename MatType, typename ResponsesType>
+  XGBoostTreeRegressor(const MatType& dataset,
+                       const ResponsesType& responses,
+                       const size_t numTrees = 100,
+                       const size_t maxDepth = 6,
+                       const double eta = 0.3,
+                       const double minChildWeight = 1,
+                       const double gamma = 0,
+                       const double lambda = 1,
+                       const double alpha = 0,
+                       const bool warmStart = false,
+                       DimensionSelectionType dimensionSelector =
+                           DimensionSelectionType());
+
+  // Train the model.
+  template<typename MatType, typename ResponsesType>
+  double Train(const MatType& dataset,
+               const ResponsesType& responses,
+               const size_t numTrees = 100,
+               const size_t maxDepth = 6,
+               const double minChildWeight = 1,
+               const double gamma = 0,
+               const double lambda = 1,
+               const double alpha = 0,
+               const bool warmStart = false,
+               DimensionSelectionType dimensionSelector =
+                   DimensionSelectionType());
+
+  template<typename VecType>
+  double Predict(const VecType& point) const;
```

Comment:
Small comment, but this should return `typename VecType::elem_type`, which may not always be `double`. :)